pkg/3scale/amp/component/backend_options.go

```diff
@@ -0,0 +1,98 @@
+package component
+
+import (
+	oprand "github.com/3scale/3scale-operator/pkg/crypto/rand"
+	"github.com/go-playground/validator/v10"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type BackendOptions struct {
+	ServiceEndpoint              string `validate:"required"`
+	RouteEndpoint                string `validate:"required"`
+	StorageURL                   string `validate:"required"`
+	QueuesURL                    string `validate:"required"`
+	StorageSentinelHosts         string
+	StorageSentinelRole          string
+	QueuesSentinelHosts          string
+	QueuesSentinelRole           string
+	ListenerResourceRequirements v1.ResourceRequirements `validate:"-"`
+	WorkerResourceRequirements   v1.ResourceRequirements `validate:"-"`
+	CronResourceRequirements     v1.ResourceRequirements `validate:"-"`
+	ListenerReplicas             int32
```

Comment:
The replicas fields are not marked as "required". This means that validating a BackendOptions where the replicas fields have not been provided would be valid. When replicas field are not provided they use the zero-value. In this case 0, which would deploy 3scale with replicas set to 0. I've seen you set those fields to 1 intentionally from the BackendOptionsProvider in both templates and operator. Is the idea giving the responsibility of creating a "coherent" "default" 3scale deployment to the operator/templates option providers implementations? (for example make sure that a 3scale with 1 replica is deployed when using defaults)

I don't have a clear idea on what's the criteria on what should be "required" and what not. What criteria will we follow to decide whether a field is "required" or not in the validator context?
To try to clarify what it was being done before, I think the criteria I tried to follow previously was that if the value was the same between templates and operator then I considered the option "non-required"(I'm talking of the old concept of required/non-required, nothing to do with the validator in this PR) as a default for it was provided in the component itself.

Reply:
The owner of the default value should be the `operator` package (or template package). It is high level information. That could be different for the operator controller and the template adapter. That does not mean that the value could not be defined in `component` package to be used by template adapters or operator controller.

Be careful with the concept of "required". Regarding the backend options, all fields are required for backend option user. For example [here](https://github.com/3scale/3scale-operator/blob/ce0b0a87de1253c478f17f70ec0f288ecc72056d/pkg/3scale/amp/component/backend.go#L252):

`Resources: *backend.Options.listenerResourceRequirements`

`listenerResourceRequirements` even if it is a pointer, it is required by the `component` package, as it is being dereferenced. Otherwise you would get Invalid memory address because of nil pointer dereference. It is correct that the value in the CRD is a pointer, because that way you differentiate between not being set and zero value. But for options being used in component, it is different.

For the options sent to the component package for building objects, generally all attributes should be required, otherwise makes no sense to pass it. Generally, there may be edge cases or more complex cases like in system component which is more complex.

Thus, required for the validator means it cannot have zero value. Back to the listener `replica`, we allow it to be 0, because that is the value set by the customer in the CR. Thus, we cannot enforce it to be not zero.

Reply:
:+1: